enhancements/automated-resource-request-scaling.md

```diff
@@ -0,0 +1,635 @@
+changes across releases. So, a new `DefaultControlPlaneCPUCount`
+constant will be added to the module with the Infrastructure type.
+
+If `controlPlaneCPULimit == 0`, operators should use their
+default configuration for their operands.
+
+If `0 < controlPlaneCPULimit < DefaultControlPlaneCPUCount`, operators
+should scale their operands down by around `(controlPlaneCPULimit /
+DefaultControlPlaneCPUCount)` or more, if appropriate. When scaling,
+the operators should continue to follow the other guidelines for
+resources and limits in
+[CONVENTIONS.md](CONVENTIONS.md#resources-and-limits).
+
+If `controlPlaneCPULimit >= DefaultControlPlaneCPUCount`, operators
+should use their default configuration for their operands. We do not
+currently have a need to allow the control plane to scale up, although
+this may change later. If we do have a future requirement to support
+that, we could remove this restriction.
+
+Users will specify the `controlPlaneCPULimit` when building a cluster
+by passing the value to the installer in `install-config.yaml`.
+
+### User Stories
+
+#### Radio Access Network (RAN) Use Case
+
+In the context of telcocommunications service providers' 5G Radio Access
+Networks, it is increasingly common to see "cloud native" implementations
+of the 5G Distributed Unit (DU) component. Due to latency constraints,
+this DU component needs to be deployed very close to the radio antenna for
+which it is responsible. In practice, this can mean running this
+component on anything from a single server at the base of a remote cell
+tower or in a datacenter-like environment serving several base stations.
+
+A hypothetical DU example is an unusually resource-intensive workload,
+requiring 20 dedicated cores, 24 GiB of RAM consumed as huge pages,
+multiple SR-IOV NICs carrying several Gbps of traffic each, and
+specialized accelerator devices. The node hosting this workload must
+run a realtime kernel, be carefully tuned to ensure low-latency
+requirements can be met, and be configured to support features like
+Precision Timing Protocol (PTP).
+
+The most constrained resource for RAN deployments is CPU, so for now
+the focus of this enhancement is scaling CPU requests.  Kubernetes
+resource requests for CPU time are measured in fractional "[Kubernetes
+CPUs](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu)". For
+bare metal, 1 Kubernetes CPU is equivalent to 1 hyperthread.
+
+Due to the resource-intensive workload, the overhead for platform
+components such as the OpenShift platform and add-on operators is
+severely restricted when compared to other uses. For example, one
+customer has allocated 4 hyperthreads for all components other than
+their own workloads. OpenShift currently requires 7.
+
+### Implementation Details/Notes/Constraints
+
+1. Add the new field to the `Infrastructure` API.
+2. Expose the `controlPlaneCPULimit` setting through the installer via
```

Comment:
And 3: all components respect this.  I think this is the problem with this design - it requires EVERY component perform the same calculation (it denormalizes something inefficiently).  If the calculation is standard, I would expect components to have to do nothing.

I would prefer an approach that centralizes the responsibility for scaling these components, not divides them out inefficiently.  This isn't about operators, this is about a set of target resource constraints.  That could be to the node, an admission plugin, or to a more general purpose solution.  I don't think this scales correctly.

Also, because this is about compressible resources, eventually we will have other constraint types besides CPU, so this already hits problems with being too CPU specific.

Reply:
Centralizing it makes a lot of sense. There are a few options below for doing that, starting on line 457. 

I like the in-cluster operator approach, because it doesn't turn resource requests into something we would need to support users modifying directly. If that isn't a big concern, then I like the "external actor" approach, since we can probably do it with an ACM policy and not have to write any scaling code at all. What do you think?

As far as other compressible resources, I think we're likely to need to be able to express each of those in their own units somehow. That implies separate API fields, and I would like to keep this enhancement focused on scaling the thing we know we need to scale today (I should update the title and summary accordingly). Are you thinking of a more general API that does not use separate fields for things like CPU and network bandwidth? Or do you foresee issues waiting to consider how we might scale things other than CPU?

Reply:
I guess another benefit of the external-actor approach is that it would let us tune things differently for different use cases, which is something @browsell brought up in a comment above.

Reply:
Also one less component consuming resources on the node as you point out below.

Reply:
Thinking about the centralization more, I'm not sure an independent actor, either inside the cluster or outside, would work.

It could change the settings of Pods from Deployments managed by the CVO without a problem. But as soon as it tried to change a workload being managed by another operator, the two things would be in contention. So if each operator doesn't support scaling on its own, then we have to do the scaling in a way that's completely transparent to the Pod definitions.

Reply:
Another argument in favor of having each operator deal with scaling is that it may need to change more than the resource requests. For example, if CPU requests are reduced and we expect that might lead to slower operation (if there is no overhead for bursting), we may need to change timeouts dynamically as well.

Reply:
An outside actor could be made to work if the operators ignored the resource requests when checking for equality -- basically every operator would have to do what we said up on pin 92 that the CVO has to do.

That complicates upgrades, though, because how does an operator ever change the resource requests for the things it is managing? Maybe we use an annotation on the Deployment to set the baseline value(s)?

Reply:
Could we turn the up-front scaling (options on line 457) around by decoupling the requests from the “applied” impact on the node? If we scale the request when updating the remaining allocatable space during pod admission to a node we preserve scheduling space for the user’s pods. In practical terms this is somewhere between the two global options (lines 593 and 602), and may meet your criteria of being transparent to the Pod definition.

As noted below this does complicate debug/analysis of the system but would decoupling requests from the applied state be beneficial for other scenarios? For example scheduling a pod on a hyperthreaded node where, for security or performance reasons, it may be necessary to exclude all code execution on the sibling CPU. If such a pod requests 1 CPU and we offline the sibling the “applied” impact is 2 CPU, a similar scaling approach (although scaled up) as could be used here.

Reply:
That sounds a lot like the option described on line 610, where kubelet scales the request before reporting the available capacity on a node. Unless there's a subtle difference that I'm missing?

Reply:
Yes, apologies, I misread that option as explicitly adding in extra capacity. Re-framing my comment then would be that having kubelet scale the requests at the node (as in the option on line 610) might meet the goal of being transparent to the Pod definition avoiding some of the thrashing scenarios and may have applications in other scenarios as well.

Reply:
No worries, I'll try to clarify the language for that one. There are a lot of options to keep straight. :-)

I think the main challenge with doing it inside kubelet is then we're carrying a fork of kubelet, at least until/unless we can convince upstream to take the feature. But, I'm not a kubelet expert so perhaps there's a way to make a plugin work?